internal/services/serviceprincipals/service_principal_claims_mapping_policy_assignment_test.go

```diff
@@ -0,0 +1,135 @@
+	mappingPolicy := acceptance.TestData{
+		ResourceName: "azuread_claims_mapping_policy.test2",
+	}
+	r := ServicePrincipalClaimsMappingPolicyAssignment{}
+
+	data.ResourceTest(t, r, []resource.TestStep{
+		{
+			Config: r.basicClaimsMappingPolicyAssignment(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
+		},
+		data.ImportStep(),
+		{
+			Config: r.updateClaimsMappingPolicyAssignment(data),
+			Check: resource.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key(
+					"claims_mapping_policy_id",
+				).MatchesOtherKey(
+					check.That(mappingPolicy.ResourceName).Key(
+						"id",
+					),
+				),
+			),
+		},
+	})
+}
+
+func (ServicePrincipalClaimsMappingPolicyAssignment) basicClaimsMappingPolicyAssignment(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+provider "azuread" {}
+
+data "azuread_application_published_app_ids" "well_known" {}
+
+resource "azuread_service_principal" "msgraph" {
+  application_id = data.azuread_application_published_app_ids.well_known.result.MicrosoftGraph
+  use_existing   = true
+}
+
+resource "azuread_claims_mapping_policy" "test" {
+  definition = [
+	  "{\"ClaimsMappingPolicy\":{\"Version\":1,\"IncludeBasicClaimSet\":\"false\",\"ClaimsSchema\": [{\"Source\":\"user\",\"ID\":\"employeeid\",\"SamlClaimType\":\"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name\",\"JwtClaimType\":\"name\"},{\"Source\":\"company\",\"ID\":\"tenantcountry\",\"SamlClaimType\":\"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country\",\"JwtClaimType\":\"country\"}]}}"
+  ]
+  description = "%[1]s"
+  display_name = "integration-%[1]s"
+}
+
+resource "azuread_claims_mapping_policy_assignment" "test" {
+  service_principal_id     = azuread_service_principal.msgraph.id
+  claims_mapping_policy_id = azuread_claims_mapping_policy.test.id
+}
+`, data.RandomString)
+}
+
+func (ServicePrincipalClaimsMappingPolicyAssignment) updateClaimsMappingPolicyAssignment(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+	provider "azuread" {}
+
+	data "azuread_application_published_app_ids" "well_known" {}
+	
+	resource "azuread_service_principal" "msgraph" {
+	  application_id = data.azuread_application_published_app_ids.well_known.result.MicrosoftGraph
+	  use_existing   = true
+	}
+	
+	resource "azuread_claims_mapping_policy" "test2" {
+	  definition = [
+		  "{\"ClaimsMappingPolicy\":{\"Version\":1,\"IncludeBasicClaimSet\":\"false\",\"ClaimsSchema\": [{\"Source\":\"user\",\"ID\":\"employeeid\",\"SamlClaimType\":\"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name\",\"JwtClaimType\":\"name\"},{\"Source\":\"company\",\"ID\":\"tenantcountry\",\"SamlClaimType\":\"http://schemas.xmlsoap.org/ws/2005/05/identity/claims/country\",\"JwtClaimType\":\"country\"}]}}"
+	  ]
+	  description = "%[1]s"
+	  display_name = "integration-%[1]s"
+	}
```

Comment:
To avoid repetition, we can substitute a config here from `ServicePrincipalClaimsMappingPolicy` (e.g. `basicClaimsMappingPolicy`)

Reply:
In principle I completely understand the need to dedupe stuff like this for future maintainability, so to make sure I understand correctly here, you want to re-use the test definition from here https://github.com/hashicorp/terraform-provider-azuread/pull/733/files#diff-673ecbaaaf4f1fe043eda4ebc250325ec41d602d5992220d3b2db8f6bf2fa1d4R42 ?

Only asking to clarify because in the `basicClaimsMappingPolicy` case it's used for the basic test while here in `updateClaimsMappingPolicyAssignment` it's used for the update test.

Should we extract out the single resource block for re-use as a constant string (or similar) or did you have some other way to re-use in mind? Do you know of other places in the tests where this has been done so we'd have an example? We want to make sure this gets updated the way you're thinking.

Reply:
Thank you for the good feedback. I updated the other concerns. I'll try to look at this one Monday.

Reply:
When working through the refactoring I found our tests were broken.
Seems they were broken even before I began to work through the suggestions.
After doing some debugging I found the `odata.ID` is `nil`
when it goes to post to msgraph.

```dlv
(dlv) locals
status = 0
policy = github.com/manicminer/hamilton/msgraph.ClaimsMappingPolicy {DirectoryObject: (*"github.com/manicminer/hamilton/msgraph.DirectoryObject")(0xc000270268), Definition: *[]string nil, Description: *string nil,...+2 more}
checkPolicyAlreadyExists = github.com/manicminer/hamilton/msgraph.(*ServicePrincipalsClient).AssignClaimsMappingPolicy.func1
(dlv) print policy.ID
*"ssssssss-ssss-ssss-ssss-ssssssssss"
(dlv) ls
> github.com/manicminer/hamilton/msgraph.(*ServicePrincipalsClient).AssignClaimsMappingPolicy() ./vendor/github.com/manicminer/hamilton/msgraph/serviceprincipals.go:371 (PC: 0xbddb07)
   366:                                 return o.Error.Match(odata.ErrorAddedObjectReferencesAlreadyExist)
   367:                         }
   368:                         return false
   369:                 }
   370:
=> 371:                 body, err := json.Marshal(DirectoryObject{ODataId: policy.ODataId})
   372:                 if err != nil {
   373:                         return status, fmt.Errorf("json.Marshal(): %v", err)
   374:                 }
   375:
   376:                 _, status, _, err = c.BaseClient.Post(ctx, PostHttpRequestInput{
```

This field is needed according to
https://docs.microsoft.com/en-us/graph/api/serviceprincipal-post-claimsmappingpolicies?view=graph-rest-1.0&tabs=http#request-body

I wonder if we missed something when we applied this suggestion
https://github.com/manicminer/hamilton/pull/147#discussion_r808657419

Any ideas on what we may have missed @manicminer?

Reply:
@manicminer I eliminated duplication with the create function for the assignment, but kept the 2nd one so simulate the update. Please let me know what you think at this point. Thank you again for your reviews.